### DISCUSSION TITLE
how to logout from wave?

### DISCUSSION THREAD

**Discussion by zillionare** [h2oai/wave#1439] [Q&A] [answered]
we have configured keycloak as auth server and all worked well but I don't know how to log user out.

I'm not sure if this is a keycloak question but after looking to keycloak's doc, I still don't find any clue. Would somebody give some help?

**Comment by mturoci** [comment]
Maybe https://github.com/h2oai/wave/discussions/1325#discussioncomment-2509569 can help?

**Reply by zillionare** [reply]
> Maybe [#1325 (reply in thread)](https://github.com/h2oai/wave/discussions/1325#discussioncomment-2509569) can help?

@mturoci I've read the link you pointed, however, I still don't know how to implement it.

does the link implies that:
1. we (the app writer) should place a ui.link somewhere on the page? If so, what's the url should it be?
2. should we (the python app) handle the click event of the ui.link? 
3. you mentioned " handle("_auth/logout", newLogoutHandler(auth, broker)) " this is in go file, what does this mean? should we touch source go file?

**Reply by henrycs** [reply]
> Maybe [#1325 (reply in thread)](https://github.com/h2oai/wave/discussions/1325#discussioncomment-2509569) can help?

I use latest wave and keycloak 10.0.2 (docker), and I call _auth/logout in myapp, I get the following error:
browser responses "we are sorry" with URL: http://192.168.100.201:8080/auth/realms/master/protocol/openid-connect/logout?id_token_hint=xxxxxxxxx&post_logout_redirect_uri=192.168.100.201%3A10101

then I update the post_logout_redirect_uri parameter to "http://192.168.100.201%3A10101/demo" (which is my app page), it works!
so, is this version of wave not compatible with keycloak 10.0.2 ? @mturoci

**Comment by mturoci** [comment]
> we (the app writer) should place a ui.link somewhere on the page? If so, what's the url should it be

Yes. The URL link should be `_auth/logout`.

> should we (the python app) handle the click event of the ui.link?

No, the Go server takes care of it. It also requires an auth cookie to be set which you cannot access from python.

> 3

See my 1st answer. 

> I get the following error: browser responses "we are sorry" with URL

Can you copy paste the whole error please? It's not clear to me what this error is.

> so, is this version of wave not compatible with keycloak 10.0.2

Please make sure to configure the `oidc-end-session-url`. If set correctly, the very exact URL param is appended to the logout request within [our logout Go endpoint](https://github.com/h2oai/wave/blob/d9092aa3cbf5466973af46e530dfb7da28019402/auth.go#L448)

Let me know if this helped guys!

**Reply by henrycs** [reply]
configurations:
"H2O_WAVE_OIDC_CLIENT_ID": "wave",
"H2O_WAVE_OIDC_CLIENT_SECRET": "1d43a43c-165a-4e83-bfd1-722dc5b1bf8a",
"H2O_WAVE_OIDC_END_SESSION_URL": "http://192.168.100.201:8080/auth/realms/master/protocol/openid-connect/logout",
"H2O_WAVE_OIDC_PROVIDER_URL": "http://192.168.100.201:8080/auth/realms/master",
"H2O_WAVE_OIDC_REDIRECT_URL": "http://192.168.100.201:10101/_auth/callback",

full error message in cloak web page: We are sorry... invalid redirect uri
and from cloak console:
09:29:36,486 WARN  [org.keycloak.events] (default task-13) type=LOGOUT_ERROR, realmId=master, clientId=null, userId=null, ipAddress=192.168.100.19, error=invalid_redirect_uri, redirect_uri=192.168.100.201:10101

full URL for cloak web page:
http://192.168.100.201:8080/auth/realms/master/protocol/openid-connect/logout?id_token_hint=eyJhbGciOiJSUzI1NiIsInR5cCIgOiAiSldUIiwia2lkIiA6ICJGeWJEbllwWmtxMUJkRnZHNDdCRmRyNldyelhIRV93UDV1Z0VXYXlRSFRJIn0.eyJleHAiOjE2NTM1NTcyNTEsImlhdCI6MTY1MzU1NzE5MSwiYXV0aF90aW1lIjoxNjUzNTU3MTkxLCJqdGkiOiIzOTBmMWU1NS1hMmIxLTRkYzctODZlYy1kNDMxMDVlZDNjZmEiLCJpc3MiOiJodHRwOi8vMTkyLjE2OC4xMDAuMjAxOjgwODAvYXV0aC9yZWFsbXMvbWFzdGVyIiwiYXVkIjoid2F2ZSIsInN1YiI6IjliYjBjOTkxLTFjZTgtNDU4OS1hYjI0LTBmYTM2ZmQ0OTkzNCIsInR5cCI6IklEIiwiYXpwIjoid2F2ZSIsIm5vbmNlIjoiOWMwMTYwOGUiLCJzZXNzaW9uX3N0YXRlIjoiMGVlNWQwYmMtM2VkMS00NjU1LTk0MzgtMDhjNDQwN2U3ZjAxIiwiYWNyIjoiMSIsImVtYWlsX3ZlcmlmaWVkIjpmYWxzZSwicHJlZmVycmVkX3VzZXJuYW1lIjoiaGVucnkifQ.bQobGvBr896r00-KufeZ8YgTWkyaOgYlOUSR5ELJRvUE0XVNJKwaHtgJXcRE4Wrq-m3RhDMMaRB0IcVZHTQNjDvE8dUbKPiSdRk9Ds7J1O77B80SBWcCtpH6FEXXulDl2h2xM-TlSCUIEEsbyQzkIPVpU6y07iq59E7j7gizxCO8qdpd95ISUxFvjsTyZWBDFgqI2jaTdO37wBNe2xVv3i1V6N23pPiugon7wXJ2NT9aND1MhmU7dSlsEvaMQHfJKEkpRv4z77UB2nUNL8uMNCoG7zGhSYZIhkLYYPg7Bi4vj6UgL6V5MwdE5R0wH9Nt_2hi9XIKpL4Udl8zFN8k5Q&post_logout_redirect_uri=192.168.100.201%3A10101

thanks.

**Reply by henrycs** [reply]
@mturoci please see my reply, thanks.

**Reply by mturoci** [reply]
H2O_WAVE_OIDC_END_SESSION_URL should be the URL you want your users to be redirected to after the logout. This is usually a sign-in page or in your case probably `http://192.168.100.201:10101/demo`

**Reply by henrycs** [reply]
I try to check "H2O_WAVE_OIDC_END_SESSION_URL" in source code, it seems it's the real logout URL of cloak service, after I change the setting to http://192.168.100.201:10101/demo, and run the app again, this time, no errors reported, but I check the sessions in cloak admin UI, no logout action executed, I can see the session in "Sessions" of the test account, I logout this session manually, and refresh my app page, the browse redirect to login page correctly.  @mturoci

**Reply by mturoci** [reply]
Please ignore what I said. You are correct, the END_SESSION_URL should be the real Keycloak logout URL (as you had before). The error you are getting is most likely due to an incomplete keycloak configuration. Can you please add `http://192.168.100.201:10101/demo` among valid redirect uris within Keycloak?

Edit: Assuming the above mentioned configuration works fine, it still may be a problem on Wave side since we only use host to create `post_logout_redirect_uri` instead of `scheme+host+route`. This means your valid redirect uri should be `192.168.100.201:10101` instead of `http://192.168.100.201:10101/demo` for now.

cc @lo5

**Reply by henrycs** [reply]
it's the first time I use Keycloak...so I just follow the wave guide to configure SSO, and use Keycloak 10.0.2 mentioned in guide, I also tried to configure Keycloak in my way, but all failed.
As I mentioned in previous reply, I use "http://192.168.100.201:10101/demo" instead of "192.168.100.201:10101" in the URL string (replace the value of post_logout_redirect_uri), it works, the session will be logged out in Keycloak.

**Reply by henrycs** [reply]
![image](https://user-images.githubusercontent.com/7357611/170487517-496b11a4-3b2c-400f-887e-84f726adcf19.png)

**Reply by henrycs** [reply]
@mturoci @lo5 I add a new configuration in conf.go (0.21.1):
type AuthConf struct {
	EndSessionURL string
	PostRedirectURL string
}

and read the env vars in main.go:
	stringVar(&auth.EndSessionURL, "oidc-end-session-url", "", "OIDC end session URL")
        stringVar(&auth.PostRedirectURL, "oidc-post-redirect-url", "", "OIDC post redirect URL")

then, I set post_logout_redirect_uri in auth.go:
	query := redirectURL.Query()
	query.Set("post_logout_redirect_uri", h.auth.conf.PostRedirectURL)
	if len(idToken) > 0 {
		// required by Okta
		// https://developer.okta.com/docs/reference/api/oidc/#logout
		query.Set("id_token_hint", idToken)
	}
	redirectURL.RawQuery = query.Encode()

finally, make build-server, replace with this new waved, it works, meet my expectation.

**Reply by mturoci** [reply]
Agree, adding the `PostRedirectURL` configuration option is the way to go as it's the most flexible and should fulfil everybody's needs. @henrycs would you might opening a PR?

**Comment by zillionare** [comment, marked as answer]
this has been fixed and available at wave 0.22.

**Reply by azim-b** [reply]
@zillionare can you point me to where in docs you found this?

**Reply by zillionare** [reply]
@azim-b the PR is provided by my colleage, and we found it's merged in 0.22.0. Not sure if it's mentioned in docs, but it's verified by us.
